### ISSUE TITLE
Fix error regarding contract creation address

### ISSUE THREAD

**Issue by jochem-brouwer** [ethereumbook/ethereumbook#963]
The book states that any transaction to the `zero address` (i.e. `0x0000..0000`) is a special transaction which creates a contract. This is wrong. A contract is created if you send it to the `null` address. You can send a transaction just fine to `0x0000...0000` which does not create a contract. 

Excerpt from the Ethereum Yellow Paper: (page 4, section 4.2 "The Transaction")

> to: The 160-bit address of the message call’s recipient or, for a contract creation transaction, ∅, used here to denote the only member of B0

I.e. the ∅ "address" is the only member of B0, the set of byte sequences of length 0.

**Comment by chanhosuh** [comment]
If you look at the raw serialized transaction, I don't think it matters whether you use the "null" address or the zero address, it serializes as the same thing so there is no difference in what the EVM receives.

**Comment by jochem-brouwer** [comment]
> If you look at the raw serialized transaction, I don't think it matters whether you use the "null" address or the zero address, it serializes as the same thing so there is no difference in what the EVM receives.

@chanhosuh Then explain to me how it is possible to send ETH to the zero address directly. Following this reasoning, it would try to create an empty contract and would thus **not** be sent to the zero address.

**Comment by chanhosuh** [comment]
You can send ETH to any address, including the zero address.  If a contract is created from the data payload, that ETH will be added to the contract account balance.  I'm not sure what the confusion is here, really.  Do you dispute my statement about the serialized transaction, which can be easily validated?

**Comment by jochem-brouwer** [comment]
> You can send ETH to any address, including the zero address. If a contract is created from the data payload, that ETH will be added to the contract account balance. I'm not sure what the confusion is here, really. Do you dispute my statement about the serialized transaction, which can be easily validated?

@chanhosuh If you validate all contract creation transactions, including those with **empty calldata**, then you will see that the address to: field is empty. You can send transactions to the zero address with calldata which still do not create a contract. This is per the Yellow Paper.

If you show me a transaction which is to: zero address and which **creates a contract** then I am convinced.

**Comment by chanhosuh** [comment]
@jochem-brouwer Yeah, I'll concede the point.  I checked several transactions that created contracts and found `0x80` there instead of an RLP-ed address.  There were several stuck at the zero address with contract bytecode as data but those didn't create contracts.

Interestingly, I noticed the Solidity docs were updated at some point to change some references from zero address to null.

**Comment by ronytesler** [comment]
So "In order to create a new contract, a special transaction is needed that has its to field set to the special 0x0 address and its data field set to the contract’s initiation code." is wrong? If it's wrong, why chapter 12 wasn't fixed until now?

**Comment by jochem-brouwer** [comment]
Is this fixed? When I opened this issue it was a problem/bug in the book.

**Comment by alessandromazza98** [comment]
> Is this fixed? When I opened this issue it was a problem/bug in the book.

new edition of the book is out now

**Comment by jochem-brouwer** [comment]
The problem is still there @alessandromazza98. In Chapter 2 it sets the `to` field of the transaction to `0x0000000000000000000000000000000000000000` in order to create a contract: https://github.com/ethereumbook/ethereumbook/blob/f1b7af44e413620344915738737d591156f84a2d/src/chapter_2.md#L417

This is not true. Trying to create a contract like this will not create a contract, it will just be a normal transaction and it will not deposit any code. It will "send" to the zero address.

In Chapter 6 this happens again, but now the definition of "zero address" has changed, now the `to` field is `0x0` (not `0x0000000000000000000000000000000000000000` as in Chapter 2), which looks "more correct", but it should be made clear to the reader that the correct value is the empty bytes (`0x`): https://github.com/ethereumbook/ethereumbook/blob/f1b7af44e413620344915738737d591156f84a2d/src/chapter_6.md#L574

It states: 

> the `to` field in a contract-registration transaction contains the address `0x0`.

It should be made clear that the `to` field is not an address but rather the empty bytes `0x`.

See also for instance Geth code: https://github.com/ethereum/go-ethereum/blob/b04df226fa142525cdc4bf0defe31bf977d35fb9/interfaces.go#L163
https://github.com/ethereum/go-ethereum/blob/b04df226fa142525cdc4bf0defe31bf977d35fb9/core/state_transition.go#L441
https://github.com/ethereum/go-ethereum/blob/b04df226fa142525cdc4bf0defe31bf977d35fb9/core/state_transition.go#L500

A bit disappointing to see that a new version of the book is being published without first going through the past issues. I hope it can still be fixed.

I'll open a new issue.

**Comment by alessandromazza98** [comment]
> I'll open a new issue.

yes please do

**Comment by jochem-brouwer** [comment]
See https://github.com/ethereumbook/ethereumbook/issues/1263
